akka-http/src/main/scala/akka/http/scaladsl/marshalling/PredefinedToResponseMarshallers.scala

```diff
@@ -79,36 +79,12 @@ trait PredefinedToResponseMarshallers extends LowPriorityToResponseMarshallerImp
         }
     })
 
-  implicit def fromEntityStreamingSupportAndByteStringMarshaller[T, M](implicit s: EntityStreamingSupport, m: ToByteStringMarshaller[T]): ToResponseMarshaller[Source[T, M]] = {
-    Marshaller[Source[T, M], HttpResponse] { implicit ec ⇒ source ⇒
-      FastFuture successful {
-        Marshalling.WithFixedContentType(s.contentType, () ⇒ {
-          val availableMarshallingsPerElement = source.mapAsync(1) { t ⇒ m(t)(ec) }
-
-          // TODO optimise such that we pick the optimal marshalling only once (headAndTail needed?)
-          // TODO, NOTE: this is somewhat duplicated from Marshal.scala it could be made DRYer
-          val bestMarshallingPerElement = availableMarshallingsPerElement mapConcat { marshallings ⇒
-            // pick the Marshalling that matches our EntityStreamingSupport
-            (s.contentType match {
-              case best @ (_: ContentType.Binary | _: ContentType.WithFixedCharset | _: ContentType.WithMissingCharset) ⇒
-                marshallings collectFirst { case Marshalling.WithFixedContentType(`best`, marshal) ⇒ marshal }
-
-              case best @ ContentType.WithCharset(bestMT, bestCS) ⇒
-                marshallings collectFirst {
-                  case Marshalling.WithFixedContentType(`best`, marshal) ⇒ marshal
-                  case Marshalling.WithOpenCharset(`bestMT`, marshal)    ⇒ () ⇒ marshal(bestCS)
-                }
-            }).toList
-          }
-          val marshalledElements: Source[ByteString, M] =
-            bestMarshallingPerElement.map(_.apply()) // marshal!
-              .via(s.framingRenderer)
+  @deprecated("This method exists only for the purpose of binary compatibility, it used to be implicit.", "10.1.0")
+  implicit def fromEntityStreamingSupportAndByteStringMarshaller[T, M](s: EntityStreamingSupport, m: ToByteStringMarshaller[T]): ToResponseMarshaller[Source[T, M]] =
+    fromEntityStreamingSupportAndByteStringMarshaller(null, s, m)
 
-          HttpResponse(entity = HttpEntity(s.contentType, marshalledElements))
-        }) :: Nil
-      }
-    }
-  }
+  implicit def fromEntityStreamingSupportAndByteStringMarshaller[T: ClassTag, M](implicit s: EntityStreamingSupport, m: ToByteStringMarshaller[T]): ToResponseMarshaller[Source[T, M]] =
+    fromEntityStreamingSupportAndByteStringSourceMarshaller(s, m.map(Source.single))
```

Comment:
Is this fine, or will wrapping each `ByteString` into a `Source.single`, only to be `mapConcat`ed immediately afterwards, add an overhead?

If it isn't fine, then we could at least commonize the "bestMarshallingPerElement" logic, and keep everything else duplicated.

Reply:
I think here it's likely fine... since it's going to materialize a stream in any case